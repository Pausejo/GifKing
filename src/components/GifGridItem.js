import React from 'react';
import PropTypes from 'prop-types';

export const GifGridItem = ({url, title}) => {

  const handleClick = () => {    
    navigator.clipboard.writeText(url);
  }

  return <div className="card" onClick={handleClick}>
            <h2 className="title">{title}</h2>
            <img src={url} alt={title} />
        </div>
}

GifGridItem.propTypes = {
  url: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
}