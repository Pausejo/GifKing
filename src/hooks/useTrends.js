import { useState, useEffect } from 'react'
import { getTrends } from '../helpers/getTrends';

export const useTrends = () => {    
    const [state, setState] = useState({
        data: [],
        loading: true,
    });

    useEffect(() => {
        getTrends()
            .then(trends => {                        
                setState({
                    data: trends,
                    loading: false,
                });
            });
    }, []); 

    return state;
}