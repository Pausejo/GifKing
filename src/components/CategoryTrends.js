import React from 'react'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faSync, faArrowTrendUp } from '@fortawesome/free-solid-svg-icons'
import { useTrends } from '../hooks/useTrends';

export const CategoryTrends = ({setCategories}) => {

    const {data: trends, loading} = useTrends();

    const trendClick = (trend) => {
        setCategories(cat => {
            return (!cat.find(c => c === trend.target.textContent)) ? 
                [trend.target.textContent, ...cat] : 
                cat
        });
    }

    return (<>
                <h3 className='trending'><FontAwesomeIcon icon={faArrowTrendUp} /> Trending</h3>
                
                <ol className="filters">
                    {
                        trends.map(t =>
                            <li key={t}>
                                <label onClick={trendClick}>{t}</label>
                            </li>)
                    }
                </ol>

                {loading && <FontAwesomeIcon icon={faSync} spin size='1x' />}
            </>
    )
}
