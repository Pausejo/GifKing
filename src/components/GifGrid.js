import React from 'react';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { faCircleMinus, faSync } from '@fortawesome/free-solid-svg-icons';
import { useFetchGifs } from '../hooks/useFetchGifs';
import { GifGridItem } from './GifGridItem';
import PropTypes from 'prop-types';

export const GifGrid = ({category, setCategories}) => {
    const {data: images, loading} = useFetchGifs(category);

    const deleteCatagory = () => {        
        setCategories((c) => c.filter(i => i!== category));
    }

    return (
         <div role="GifGrid">
            <h2><label onClick={deleteCatagory}><FontAwesomeIcon className="delete" icon={faCircleMinus} /> {category}</label></h2>        

            {loading && <FontAwesomeIcon data-testid="loading" icon={faSync} spin size='4x' />}

            <div className="container">
                { images.map(img => <GifGridItem key={img.id} {...img} />) }            
            </div>
        </div>
    )
}

GifGrid.propTypes = {
    category: PropTypes.string.isRequired,
    setCategories: PropTypes.func.isRequired,
}