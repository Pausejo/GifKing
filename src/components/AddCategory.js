import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { CategoryTrends } from './CategoryTrends';

export const AddCategory = ({setCategories}) => {

    const [inputValue, setInputValue] = useState('');

    const handleInputChange = (e)=>{
        setInputValue(e.target.value);        
    }

    const handleSubmit = (e)=>{
        e.preventDefault();

        if (inputValue.length>2) {
            setCategories(cat => {
                return (!cat.find(c => c === inputValue)) ? 
                    [inputValue, ...cat] : 
                    cat
            });
            setInputValue('');
        }
    }

    return (<>
            <form data-testid="form" onSubmit={handleSubmit}>
                <div className="row">
                    <input
                        className="balloon"
                        id="category"
                        type="text"
                        value={inputValue}
                        onChange={handleInputChange}
                        placeholder="Ingrese la categorÃ­a" />
                        <label htmlFor="category">Category</label>
                </div>
            </form>
            <CategoryTrends setCategories={setCategories}/>
        </>
    );
}

AddCategory.propTypes = {
    setCategories: PropTypes.func.isRequired,
}