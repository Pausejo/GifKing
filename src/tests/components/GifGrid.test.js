import { render, screen } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs');

describe('Testing GifGrid', () => { 

    const setCategories=jest.fn();
    const category = 'Viyuela';

    test('should show component correctly', () => { 

        useFetchGifs.mockReturnValue({
            data: [],
            loading: true,
        });

        const tree = renderer.create(<GifGrid category={category} setCategories={setCategories}/>).toJSON();

        expect(tree).toMatchSnapshot(); 
    });

    test('should show items when images are loaded', () => { 
        
        const gifs = [{
            id: '1',
            url: 'https://mockiando.com/test.gif',
            title: 'Mock'
        },{
            id: '2',
            url: 'https://mockiando.com/test2.gif',
            title: 'Mock2'
        }];

        useFetchGifs.mockReturnValue({
            data: gifs,
            loading: false,
        });

        render(<GifGrid category={category} setCategories={setCategories}/>);

        expect(screen.queryByTestId("loading")).toBeNull();
        
        expect(screen.getAllByRole("img").length).toBe(gifs.length);

    });
 })