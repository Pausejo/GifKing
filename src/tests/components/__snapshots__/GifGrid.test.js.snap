// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing GifGrid should show component correctly 1`] = `
<div
  role="GifGrid"
>
  <h2>
    <label
      onClick={[Function]}
    >
      <svg
        aria-hidden="true"
        className="svg-inline--fa fa-circle-minus delete"
        data-icon="circle-minus"
        data-prefix="fas"
        focusable="false"
        role="img"
        style={Object {}}
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M0 256C0 114.6 114.6 0 256 0C397.4 0 512 114.6 512 256C512 397.4 397.4 512 256 512C114.6 512 0 397.4 0 256zM168 232C154.7 232 144 242.7 144 256C144 269.3 154.7 280 168 280H344C357.3 280 368 269.3 368 256C368 242.7 357.3 232 344 232H168z"
          fill="currentColor"
          style={Object {}}
        />
      </svg>
       
      Viyuela
    </label>
  </h2>
  <svg
    aria-hidden="true"
    className="svg-inline--fa fa-arrows-rotate fa-spin fa-4x "
    data-icon="arrows-rotate"
    data-prefix="fas"
    data-testid="loading"
    focusable="false"
    role="img"
    style={Object {}}
    viewBox="0 0 512 512"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M464 16c-17.67 0-32 14.31-32 32v74.09C392.1 66.52 327.4 32 256 32C161.5 32 78.59 92.34 49.58 182.2c-5.438 16.81 3.797 34.88 20.61 40.28c16.89 5.5 34.88-3.812 40.3-20.59C130.9 138.5 189.4 96 256 96c50.5 0 96.26 24.55 124.4 64H336c-17.67 0-32 14.31-32 32s14.33 32 32 32h128c17.67 0 32-14.31 32-32V48C496 30.31 481.7 16 464 16zM441.8 289.6c-16.92-5.438-34.88 3.812-40.3 20.59C381.1 373.5 322.6 416 256 416c-50.5 0-96.25-24.55-124.4-64H176c17.67 0 32-14.31 32-32s-14.33-32-32-32h-128c-17.67 0-32 14.31-32 32v144c0 17.69 14.33 32 32 32s32-14.31 32-32v-74.09C119.9 445.5 184.6 480 255.1 480c94.45 0 177.4-60.34 206.4-150.2C467.9 313 458.6 294.1 441.8 289.6z"
      fill="currentColor"
      style={Object {}}
    />
  </svg>
  <div
    className="container"
  />
</div>
`;
